python/test/unit/language/test_core.py

```diff
@@ -380,28 +387,35 @@ def get_scalar(x, dtype, low, high, filter):
         do_test(x, y[:1].reshape(()), kernel_broadcast_rhs)
 
 
-def _mod_operation_ill_conditioned(dtype_x, dtype_y) -> bool:
-    # FIXME For large x, we are casting x to a floating point where it does not fit
-    #       For small y, we are computing floor(div(float(x), y)) which may not fit
-    return (dtype_x, dtype_y) in [
-        ('int32', 'bfloat16'),
-        ('int32', 'float16'),
-        ('int32', 'float32'),
-        ('int64', 'bfloat16'),
-        ('int64', 'float16'),
-        ('int64', 'float32'),
-        ('int64', 'float64'),
-        ('uint16', 'bfloat16'),
-        ('uint16', 'float16'),
-        ('uint16', 'float32'),
-        ('uint32', 'bfloat16'),
-        ('uint32', 'float16'),
-        ('uint32', 'float32'),
-        ('uint64', 'bfloat16'),
-        ('uint64', 'float16'),
-        ('uint64', 'float32'),
-        ('uint64', 'float64'),
-    ]
+def _min_max_integral_mod_value(dtype_x, dtype_y) -> Optional[int]:
+    """
+    Limit min/max values for integral types for mod values. Leads to
+    overflow/underflow when casting large integral types to floats.
+    """
+    x_bitwidth = _bitwidth(dtype_x)
+    y_bitwidth = _bitwidth(dtype_y)
+
+    # uint types have larger range, must limit min/max values
+    if x_bitwidth < y_bitwidth and dtype_x in int_dtypes:
+        return None, None
+
+    # Limit max value bit-width to be one integral type less than y-bitwidth
+    # For example:
+    #   int64, float32 -> int16
+    #   uint16, float16 -> uint8
+    #   int8 -> int8 (no int4 type in numpy)
+    x_dtype = _dtype(dtype_x)
+    max_bitwidth = max(x_bitwidth >> (x_bitwidth // y_bitwidth), 8)
+    dtype_max = x_dtype + str(max_bitwidth)
+
+    max_info = np.iinfo(getattr(np, dtype_max))
+
+    if max_bitwidth == 8:
+        return max_info.min, max_info.max
+    elif dtype_x in uint_dtypes:
+        return max_info.min, max_info.max // 16
+    else:
+        return max_info.min // 8, max_info.max // 8
```

Comment:
a LUT for this is also an option. Not a big fan of this implementation, 16 and 8 were the two lowest power of two that made the values not overflow/underflow on my local.